### PR TITLE
feature(zzh): add STEVE algorithm

### PR DESCRIPTION
**Pull request by ZHZisZZ**
## Description
add `STEVESACPolicy` in `ding/policy/mbpolicy/mbsac.py`

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [x] pass style check
- [x] pass all the tests

### CODE REVIEW

File: ding/policy/mbpolicy/mbsac.py [hunk 1]
```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
             'alpha',
             'target_q_value',
         ] + alpha_loss
+
+
+@POLICY_REGISTRY.register('stevesac')
+class STEVESACPolicy(SACPolicy):
```
Comment (jiaruonan) [inline ding/policy/mbpolicy/mbsac.py:227]: Origin STEVE uses DDPG as policy improvement, so we should replace the sac?

Comment (ZHZisZZ) [reply ding/policy/mbpolicy/mbsac.py:227]: STEVE uses DDPG because its baseline algorithm MVE uses DDPG, which is published before the SAC. Since STEVE is no longer SOTA, a complete reimplementation is not necessary. But the idea of Stochastic Ensemble Value Expansion is worth integrating into our model-based toolbox.

File: ding/policy/mbpolicy/mbsac.py [hunk 2]
```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
...
+        cum_rewards = (rewards * discounts[:-1]).cumsum(dim=0)
+        discounted_q_values = target_q_values * discounts[1:]
+        steve_return = cum_rewards + discounted_q_values
+        # (T, B)
+        steve_return_mean = steve_return.mean(1)
+        with torch.no_grad():
+            steve_return_inv_var = 1 / (1e-8 + steve_return.var(1, unbiased=False))
+            steve_return_weight = steve_return_inv_var / (1e-8 + steve_return_inv_var.sum(dim=0))
+        # (B)
+        steve_return = (steve_return_mean * steve_return_weight).sum(0)
```
Comment (jiaruonan) [inline ding/policy/mbpolicy/mbsac.py:343]: I tested the program logic of calculating STEVE target return. Well done! 
Compared with the origin paper, we didn't have an additional ensemble of reward model and followed MBPO ensemble form. So the ensemble calculation is clear.

Comment (ZHZisZZ) [reply ding/policy/mbpolicy/mbsac.py:343]: We do have an ensemble of reward models but they are actually coupled with the dynamics model. 
We do not have a Q ensemble and we also do not plan to have a Q ensemble.

File: ding/torch_utils/reshape_helper.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+from typing import Tuple, Union
```
Comment (PaParaZz1) [inline ding/torch_utils/reshape_helper.py:1]: add unittest

File: ding/torch_utils/reshape_helper.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+from torch import Tensor, Size
+
+
+def fold_batch(x: Tensor, nonbatch_dims=1) -> Tuple[Tensor, Size]:
```
Comment (PaParaZz1) [inline ding/torch_utils/reshape_helper.py:6]: why use `dims` to refer to a integer

File: ding/world_model/base_world_model.py [hunk 5]
```diff
@@ -215,10 +215,12 @@ def step(obs, act):
...
             timesteps = {
                 id: BaseEnvTimestep(n, r, d, {})
-                for id, n, r, d in zip(data_id,
-                                       next_obs.cpu().numpy(),
-                                       rewards.cpu().numpy(),
-                                       terminals.cpu().numpy())
+                for id, n, r, d in zip(
+                    data_id,
+                    next_obs.cpu().numpy(),
+                    rewards.unsqueeze(-1).cpu().numpy(),  # ding api
```
Comment (PaParaZz1) [inline ding/world_model/base_world_model.py:221]: can we define reward.unsqueeze in model-based env, just like other real envs

File: ding/world_model/base_world_model.py [hunk 6]
```diff
@@ -334,9 +331,9 @@ def rollout(self, obs: Tensor, actor_fn: Callable[[Tensor], Tuple[Tensor, Tensor
             torch.stack(obss),
             torch.stack(actions),
             # rewards is an empty list when horizon=0
-            torch.stack(rewards) if rewards else torch.tensor(rewards, device=device),
+            torch.stack(rewards) if rewards else torch.tensor(rewards, device=obs.device),
             torch.stack(aug_rewards),
-            torch.stack(dones)
+            torch.stack(dones) if dones else torch.tensor(dones, device=obs.device)
```
Comment (PaParaZz1) [inline ding/world_model/base_world_model.py:336]: maybe reward.device can be more general, because obs often owns diverse structures

File: ding/world_model/base_world_model.py [hunk 7]
```diff
@@ -334,9 +331,9 @@ def rollout(self, obs: Tensor, actor_fn: Callable[[Tensor], Tuple[Tensor, Tensor
             torch.stack(obss),
             torch.stack(actions),
             # rewards is an empty list when horizon=0
-            torch.stack(rewards) if rewards else torch.tensor(rewards, device=device),
+            torch.stack(rewards) if rewards else torch.tensor(rewards, device=obs.device),
```
Comment (PaParaZz1) [inline ding/world_model/base_world_model.py:334]: modify it to a more clear condition, `if xxx` can be a little confuse here

File: ding/policy/mbpolicy/mbsac.py [hunk 8]
```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
...
+            use_nstep=False
+        )
+        if self._cuda:
+            data = to_device(data, self._device)
+
+        if len(data['action'].shape) == 1:
+            data['action'] = data['action'].unsqueeze(1)
+
+        # [B, D] -> [E, B, D]
+        data['next_obs'] = data['next_obs'].unsqueeze(0).repeat(self._ensemble_size, 1, 1)
```
Comment (PaParaZz1) [inline ding/policy/mbpolicy/mbsac.py:311]: maybe a `repeat_batch` here

File: ding/policy/mbpolicy/mbsac.py [hunk 9]
```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
...
+        self._auto_alpha = False
+
+        def actor_fn(obs: Tensor):
+            obs, dim = fold_batch(obs, 1)
+            (mu, sigma) = self._learn_model.forward(obs, mode='compute_actor')['logit']
+            dist = Independent(Normal(mu, sigma), 1)
+            pred = dist.rsample()
+            action = torch.tanh(pred)
+
+            log_prob = dist.log_prob(
```
Comment (PaParaZz1) [inline ding/policy/mbpolicy/mbsac.py:279]: why do this operation here, if it aims to correct tanh, why not use the implementation [here](https://github.com/opendilab/DI-engine/blob/main/ding/policy/sac.py#L819)

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/torch_utils/reshape_helper.py (hunk 3 above)
Comment (ZHZisZZ) [reply ding/torch_utils/reshape_helper.py:1]: added

File: ding/torch_utils/reshape_helper.py (hunk 4 above)
Comment (ZHZisZZ) [reply ding/torch_utils/reshape_helper.py:6]: changed to `ndims` following `torch.distributions.Independent`.

File: ding/policy/mbpolicy/mbsac.py [hunk 10]
```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
...
+        # (T, E, B)
+        cum_rewards = (rewards * discounts[:-1]).cumsum(dim=0)
+        discounted_q_values = target_q_values * discounts[1:]
+        steve_return = cum_rewards + discounted_q_values
+        # (T, B)
+        steve_return_mean = steve_return.mean(1)
+        with torch.no_grad():
+            steve_return_inv_var = 1 / (1e-8 + steve_return.var(1, unbiased=False))
+            steve_return_weight = steve_return_inv_var / (1e-8 + steve_return_inv_var.sum(dim=0))
+        # (B)
```
[the remark this answers is not in the document]
Comment (ZHZisZZ) [reply ding/policy/mbpolicy/mbsac.py:342]: changed

File: ding/policy/mbpolicy/mbsac.py [hunk 11]
```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
...
+    def _init_learn(self) -> None:
+        super()._init_learn()
+
+        self._grad_clip = self._cfg.learn.grad_clip
+        self._ensemble_size = self._cfg.learn.ensemble_size
+
+        self._target_model.requires_grad_(False)
+
+        # TODO: auto alpha
+        self._auto_alpha = False
```
[the remark this answers is not in the document]
Comment (ZHZisZZ) [reply ding/policy/mbpolicy/mbsac.py:270]: changed

File: ding/world_model/base_world_model.py (hunk 7 above)
Comment (ZHZisZZ) [reply ding/world_model/base_world_model.py:334]: I can change to ` torch.stack(rewards) if rewards != [] else torch.tensor(rewards, device=obs.device)`, but I do not think it is necessary.

File: ding/world_model/base_world_model.py (hunk 6 above)
Comment (ZHZisZZ) [reply ding/world_model/base_world_model.py:336]: I can not change to `reward.device`, because MBRL starts only with `obs` and then do rollouts. And in MBRL, I do not think it is common that `obs` will own diverse structures because in `DreamWorldModel`, `obs` have to be an output from networks.

File: ding/policy/mbpolicy/mbsac.py (hunk 8 above)
Comment (ZHZisZZ) [reply ding/policy/mbpolicy/mbsac.py:311]: Is `repeat_batch` a `pytorch` funtion?

File: ding/world_model/base_world_model.py (hunk 5 above)
Comment (ZHZisZZ) [reply ding/world_model/base_world_model.py:221]: The short answer is maybe we can't.
`(B, 1)` size is only necessary when `WorldModel` has to interact with `IBuffer`. The data fetched from the buffer is then passed to `default_collate` to squeeze the trailing dimension to be `(B,)`.
However, in some other cases, as in `mbsac`, `WorldModel` interacts with the policy directly and produces batched data (since transitions are now outputs from the neurel nets) so the best output size is `(B, )`, otherwise I need to squeeze here and there in the policy side, which is redundant.

File: ding/policy/mbpolicy/mbsac.py (hunk 9 above)
Comment (PaParaZz1) [reply ding/policy/mbpolicy/mbsac.py:279]: ok, we need to unity this implementation in all the codes related to SAC in the future

File: ding/policy/mbpolicy/mbsac.py (hunk 8 above)
Comment (PaParaZz1) [reply ding/policy/mbpolicy/mbsac.py:311]: no, you need to create a function like `fold_batch`

File: ding/world_model/base_world_model.py (hunk 7 above)
Comment (PaParaZz1) [reply ding/world_model/base_world_model.py:334]: ok, just add more comments about this

File: ding/world_model/base_world_model.py (hunk 5 above)
Comment (PaParaZz1) [reply ding/world_model/base_world_model.py:221]: okk

File: ding/world_model/base_world_model.py (hunk 7 above)
Comment (ZHZisZZ) [reply ding/world_model/base_world_model.py:334]: sure

File: ding/policy/mbpolicy/mbsac.py (hunk 8 above)
Comment (ZHZisZZ) [reply ding/policy/mbpolicy/mbsac.py]: I create a helper function `unsqueeze_repeat`

**Review by PaParaZz1 — APPROVED** [review]
